src/v/arch/manifest.h

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright 2021 Vectorized, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/vectorizedio/redpanda/blob/master/licenses/rcl.md
+ */
+
+#pragma once
+
+#include "bytes/iobuf.h"
+#include "model/fundamental.h"
+#include "model/metadata.h"
+#include "seastarx.h"
+#include "storage/ntp_config.h"
+
+#include <boost/property_tree/ptree_fwd.hpp>
+
+#include <iterator>
+
+namespace arch {
+
+/// Segment file name without working directory,
+/// expected format: <base-offset>-<term-id>-<revision>.log
+using segment_name = named_type<ss::sstring, struct arch_segment_name_t>;
+/// Segment path in S3, expected format:
+/// <prefix>/<ns>/<topic>/<part-id>_<rev>/<base-offset>-<term-id>-<revision>.log
+using remote_segment_path
+  = named_type<ss::sstring, struct arch_remote_segment_path_t>;
+using remote_manifest_path
+  = named_type<ss::sstring, struct arch_remote_manifest_path_t>;
+/// Local segment path, expected format:
+/// <work-dir>/<ns>/<topic>/<part-id>_<rev>/<base-offset>-<term-id>-<revision>.log
+using local_segment_path
+  = named_type<ss::sstring, struct arch_local_segment_path_t>;
+
+/// Manifest file stored in S3
+class manifest final {
+    using ptree = boost::property_tree::ptree;
+
+public:
+    struct segment_meta {
+        bool is_compacted;
+        size_t size_bytes;
+        model::offset base_offset;
+        model::offset committed_offset;
+        /// Set to true if the file was deleted
+        bool is_deleted_locally; // NOTE: because S3 doesn't delete files
```

Comment:
beautiful. up to you if you want to use a `ss::bool_class<>` if you want a strong type.